### PR TITLE
[WIP] Check in initial IMDSv2 migration work

### PR DESCRIPTION
**Pull request by zackproser**
This is a work in progress, so feel free to ignore for the time being. 

TODO: 

- [ ] Update test harness to support IMDSv2 token functionality 
- [ ] Complete `bash-commons` module IMDSv2 migration 
- [ ] Update documentation and comments

### CODE REVIEW

File: modules/bash-commons/src/aws.sh [hunk 1]
```diff
@@ -3,26 +3,200 @@
...
+# This function has been modified to simultaneously support Instance Metadata service versions 1 and 2
+# This is due to the fact that we will need to operate in a split-brain mode while all our dependent
+# modules are being updated to use IMDSv2.
+#
+# Version 2 is now the default used by this library, but the  version to use can be overridden by setting
+# env var GRUNTWORK_BASH_COMMONS_IMDSV=1
 function aws_lookup_path_in_instance_metadata {
   local -r path="$1"
-  curl --silent --show-error --location "http://169.254.169.254/latest/meta-data/$path/"
+    version_in_use=$(aws_get_instance_metadata_version_in_use)
```
Comment (bwhaley) [inline modules/bash-commons/src/aws.sh:56]:
```suggestion
  version_in_use=$(aws_get_instance_metadata_version_in_use)
```

Comment (bwhaley) [reply modules/bash-commons/src/aws.sh:56]: Indentation appears to be slightly off.

File: modules/bash-commons/src/aws.sh [hunk 2]
```diff
@@ -3,26 +3,200 @@
...
+  local -r field="$1"
+  local ttl=""
+  ttl=$(configure_imdsv2_ttl "$2")
+  token=$(ec2_metadata_http_put "$ttl")
+  curl "$instance_identity_endpoint" -H "X-aws-ec2-metadata-token: $token" \
+    --silent --location --fail --show-error | jq -r ".${field}"
+}
+
+# This is a convenience function for setting a TTL, and falling back to sensible defaults
+# when the value is either not supplied or out of bounds.
```
Comment (bwhaley) [inline modules/bash-commons/src/aws.sh:165]: Should the behavior be to throw an error when out of bounds instead of falling back to the default? If we just ignore the invalid value and move on, a user might mistakenly believe that an invalid value is valid.

**Review by bwhaley — COMMENTED** [review]
Great to see this. I wasn't familiar with the IMDS v2 updates, but after reading this I think I mostly get it.

I am a little confused about backward compatibility. I understand the split function solution, and I see that there is a check for which version based on `GRUNTWORK_BASH_COMMONS_IMDS_VERSION` and falling back to the default. But the default is 2, not 1, so the caller will have to explicitly set `GRUNTWORK_BASH_COMMONS_IMDS_VERSION=1`. Isn't that backward incompatible?

> Token caching behavior (not currently implemented - thoughts on if this is necessary / how to best support it very welcome)

I'd actually turn it back at you as you are now the IMDS v2 expert! Can you think of situations in which caching would be needed? Superficially, it seems like such a cheap call to get a token that you could just make it on demand before each request, but maybe that's naive.

> Right now users can either specify a TTL when calling the new IMDS V2 functions - otherwise the max session default (6 hours) is used

Seems reasonable. What's the max valid period of a token according to AWS? Should we set the max to their max?

> Additional test coverage - which specific tests cases should be added?

Hmm, not sure. You've already covered all the same tests as v1. Perhaps add tests with unexpected TTL values (normal value, negative value, unusually high value)?

Looks like the tests here are mocks.. would it be worthwhile to test on a "real" instance?

**Comment by zackproser** [comment]
Thanks for reviewing!

> I am a little confused about backward compatibility. I understand the split function solution, and I see that there is a check for which version based on `GRUNTWORK_BASH_COMMONS_IMDS_VERSION` and falling back to the default. But the default is 2, not 1, so the caller will have to explicitly set `GRUNTWORK_BASH_COMMONS_IMDS_VERSION=1`. Isn't that backward incompatible?

Yes, good question. If I've done this correctly, the intention is that the function calls are entirely backward compatible. This is because the arity of the original calls has been preserved, and the calls that use IMDSv2 are handling their token retrieval and presentation on their own, but returning the exact same output for the exact same input as in version 1. 

That said, I implemented the switch because we do have to operate in a split-brain mode while we're updating all our modules to use IMDSv2. Once we're satisified that everything is cut over to v2 without issue, we wanted to use version 2 as our most secure default going forward, while leaving an escape valve for anyone who might want or need to use version 1.  

> Token caching behavior (not currently implemented - thoughts on if this is necessary / how to best support it very welcome) 
> I'd actually turn it back at you as you are now the IMDS v2 expert! Can you think of situations in which caching would be needed? Superficially, it seems like such a cheap call to get a token that you could just make it on demand before each request, but maybe that's naive.

It is indeed a cheap call, and even [the official AWS docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html) state that it is perfectly fine to retrieve a new token for each transaction as there are no limits to the number of concurrent sessions or tokens that can be issued for a given instance. 

That said, the IMDS is still subject to the default throttling limits for AWS endpoints. I haven't yet identified a use-case that would need to constantly retrieve data from IMDS at sufficiently high frequency to risk tripping the rate limit. Any such use cases would theoretically be arguments for a first-class caching mechanism. 

> Right now users can either specify a TTL when calling the new IMDS V2 functions - otherwise the max session default (6 hours) is used

> Seems reasonable. What's the max valid period of a token according to AWS? Should we set the max to their max?

The max session duration for an IMDSv2 token is 6 hours. 

> > Additional test coverage - which specific tests cases should be added?
 
> Hmm, not sure. You've already covered all the same tests as v1. Perhaps add tests with unexpected TTL values (normal value, negative value, unusually high value)?

Yep, good idea!

> Looks like the tests here are mocks.. would it be worthwhile to test on a "real" instance?

It probably would, though I've extensively testing all of these on instances as I go. Doesn't mean there couldn't or shouldn't be an end to end ec2 example - I may initially file that as a follow-up ticket.

**Comment by bwhaley** [comment]
> That said, I implemented the switch because **we do have to operate in a split-brain mode while we're updating all our modules to use IMDSv2**. Once we're satisified that everything is cut over to v2 without issue, we wanted to use version 2 as our most secure default going forward, while leaving an escape valve for anyone who might want or need to use version 1.

I think I'm still missing something. The bolded part is what I'm confused about. If the calls are "entirely backward compatible," then why do we have to maintain split-brain while we're updating other modules? And if they are _not_ entirely backward compatible, then setting `GRUNTWORK_BASH_COMMONS_IMDS_VERSION=1` will retain the original behavior. Is the idea to go to all the modules that still need v1, update them to set that env var explicitly, then come back again and update them for IMDS v2 sometime later?

> That said, the IMDS is still subject to the default throttling limits for AWS endpoints. I haven't yet identified a use-case that would need to constantly retrieve data from IMDS at sufficiently high frequency to risk tripping the rate limit. Any such use cases would theoretically be arguments for a first-class caching mechanism.

Given that we haven't identified a use case, IMO we shouldn't implement caching yet. We can always do so later if the need arises.

> It probably would, though I've extensively testing all of these on instances as I go. Doesn't mean there couldn't or shouldn't be an end to end ec2 example - I may initially file that as a follow-up ticket.

Sound good.

**Comment by zackproser** [comment]
> I think I'm still missing something. The bolded part is what I'm confused about. If the calls are "entirely backward compatible," then why do we have to maintain split-brain while we're updating other modules? And if they are _not_ entirely backward compatible, then setting `GRUNTWORK_BASH_COMMONS_IMDS_VERSION=1` will retain the original behavior. Is the idea to go to all the modules that still need v1, update them to set that env var explicitly, then come back again and update them for IMDS v2 sometime later?

Ah, gotcha. Sorry - that was confusing as originally written. From our internal discussion about this work: 

> I think we will have to have nodes support both versions while all the modules are being updated, but once we confirm all our tools/modules support IMDSv2, it seems like we should take the stance that only IMDSv2 should be used at that point.

> +1. Once our modules support IMDSv2, we should disable v1 and enable only v2 by default. We may still want some "escape valve" for customers who want to explicitly re-enable v1, but it doesn't need to be easy or first class.

So, I suppose I've got it backward in that we should be pinning to V1 out of caution, initially, while rolling the changes out  to consuming modules that can themselves specify version 2 can be used. 

I implemented the switch like this so that: 

1. We can use it to handle this roll out more easily 
2. Users in the future can use it to revert to version 1 (once 2 has been updated to be the default version in bash-commons) in case they have a specific need to do so 

> Given that we haven't identified a use case, IMO we shouldn't implement caching yet. We can always do so later if the need arises.

Makes sense to me :+1:

File: modules/bash-commons/src/aws.sh (hunk 2 above)
Comment (zackproser) [reply modules/bash-commons/src/aws.sh:165]: I added an error message in 89c1c88 - so the behavior is now: if a caller supplies a TTL that is greater than the 6 hour limit for the token, a warning error message will be printed explaining that the value is excessive and that the fallback of 6 hours is being used. 

I think this is a decent middle road - if we add an explicit exit this becomes much harder to test with the current harness.

**Comment by zackproser** [comment]
> LGTM. Nice work! ship

Thank you kindly for all the thoughtful reviews!

**Comment by brikis98** [comment]
Glad to see this merged!

Sanity check: do we have an issue to track that the default IMDS version should be updated to `2` in the future?

**Comment by zackproser** [comment]
Thanks @brikis98 - I captured this in https://github.com/gruntwork-io/bash-commons/issues/40 :+1:

**Comment by zackproser** [comment]
Fixes https://github.com/gruntwork-io/company/issues/132
